src/mlpack/methods/ann/layer/bilinear_interpolation.hpp

```diff
@@ -0,0 +1,159 @@
+   * @param output The resulting interpolated output matrix.
+   */
+  template<typename eT>
+  void Forward(const arma::Mat<eT>&& input, arma::Mat<eT>&& output);
+
+  /**
+   * Ordinary feed backward pass of a neural network, calculating the function
+   * f(x) by propagating x backwards through f. Using the results from the feed
+   * forward pass. Since the layer does not have any learn-able parameters,
+   * we just have down-sample the gradient to make its size compatible with
+   * the input size.
+   *
+   * @param input The input matrix.
+   * @param gradient The computed backward gradient.
+   * @param output The resulting down-sampled output image.
+   */
+  template<typename eT>
+  void Backward(const arma::Mat<eT>&& /*input*/,
+                arma::Mat<eT>&& gradient,
+                arma::Mat<eT>&& output);
+
+  /**
+   * Calculate the gradient using the output delta and the input activation.
+   *
+   * @param input The propagated input.
+   * @param error The calculated error.
+   * @param gradient The calculated gradient.
+   */
+  template<typename eT>
+  void Gradient(const arma::Mat<eT>&& /* input */,
+                arma::Mat<eT>&& error,
+                arma::Mat<eT>&& gradient);
+
+  //! Get the input parameter.
+  InputDataType const& InputParameter() const { return inputParameter; }
+  //! Modify the input parameter.
+  InputDataType& InputParameter() { return inputParameter; }
+
+  //! Get the output parameter.
+  OutputDataType const& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  //! Get the parameters.
+  OutputDataType const& Parameters() const { return weights; }
+  //! Modify the parameters.
+  OutputDataType& Parameters() { return weights; }
+
+  //! Get the gradient.
+  OutputDataType const& Gradient() const { return gradient; }
+  //! Modify the gradient.
+  OutputDataType& Gradient() { return gradient; }
+
+  //! Get the delta.
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  /**
+   * Serialize the layer.
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  //! Locally stored row size of the input.
+  size_t inRowSize;
+  //! Locally stored column size of the input.
+  size_t inColSize;
+  //! Locally stored row size of the output.
+  size_t outRowSize;
+  //! Locally stored column size of the input.
+  size_t outColSize;
+  //! Locally stored depth of the input.
+  size_t depth;
+  //! Locally-stored weight object.
+  OutputDataType weights;
+  //! Locally-stored gradient object.
+  OutputDataType gradient;
+  //! Locally-stored delta object.
+  OutputDataType delta;
+  //! Locally-stored input parameter object.
+  InputDataType inputParameter;
+  //! Locally-stored output parameter object.
+  OutputDataType outputParameter;
```

Comment:
Correct if I'm mistaken, but looks like the three variables above aren't used.

Reply:
You're correct, actually these came up while I was trying to add `CheckGradient()` function to the tests, however, that is not needed and I'll remove them.